### PR TITLE
[JENKINS-64091] Inspect commit message in PatchSetCreated trigger

### PR DESCRIPTION
**Pull request by bkhouri**
* Add support for "Commit message contains regular expression" in the
  Patch Set Created trigger
* Commit message ignored is the textbox is empty

Issue JENKINS-64091

### CODE REVIEW

File: src/main/webapp/trigger/help-CommitMessageContains_ja.html [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<p>
```
Comment (bkhouri) [inline src/main/webapp/trigger/help-CommitMessageContains_ja.html:1]: This is not Japanese, but I wasn't sure what happens if this files does not exists, and the local is JA.  So I created it and included the English text.

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/event/PluginPatchsetCreatedEventTest.java [hunk 2]
```diff
@@ -85,4 +86,48 @@ public void shouldNotFireOnNoCodeChangeWhenExcluded() {
         patchsetCreated.getPatchSet().setKind(GerritChangeKind.NO_CODE_CHANGE);
         assertFalse(pluginPatchsetCreatedEvent.shouldTriggerOn(patchsetCreated));
     }
+
+    /**
+     * Test that it should, or should not, fire if the commit message matches a regular expression.
+     */
+    @Test
+    public void shouldNotFireIfCommitMessageRegExIsNull() {
```
Comment (bkhouri) [inline src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/event/PluginPatchsetCreatedEventTest.java:94]: I am in favour of splitting this into multiple tests, however I followed the same "structure" as other tests, where they perform multiple checks.

We should either rename this method, or split into multiple tests.  Thoughts?

File: src/main/webapp/trigger/help-CommitMessageContains_ja.html (hunk 1 above)
Comment (rsandell) [reply src/main/webapp/trigger/help-CommitMessageContains_ja.html:1]: You can remove it. If there is no file for the current language then the default text, from the help file without a suffix, will be used.

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/events/PluginPatchsetCreatedEvent.java [hunk 3]
```diff
@@ -203,6 +221,10 @@ public boolean shouldTriggerOn(GerritTriggeredEvent event) {
         if (excludeWipState && ((PatchsetCreated)event).getChange().isWip()) {
             return false;
         }
+        if (null != this.commitMessageContainsRegEx && !this.commitMessageContainsRegEx.isEmpty()) {
```
Comment (rsandell) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/events/PluginPatchsetCreatedEvent.java:224]:
```suggestion
        if (StringUtils.isNotEmpty(commitMessageContainsRegEx)) {
```

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/events/PluginPatchsetCreatedEvent.java [hunk 4]
```diff
@@ -203,6 +221,10 @@ public boolean shouldTriggerOn(GerritTriggeredEvent event) {
...
+            Pattern pattern = Pattern.compile(this.commitMessageContainsRegEx, Pattern.DOTALL | Pattern.MULTILINE);
```
Comment (rsandell) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/events/PluginPatchsetCreatedEvent.java:225]: This method can potentially be called many times per second. Better to cache it in a `transient` field and only compile a new instance if the field is `null`.

File: src/main/webapp/trigger/help-CommitMessageContains.html [hunk 5]
```diff
@@ -0,0 +1,8 @@
+<p>
+	Trigger Build if commit message matches a regular expression.
+</p>
+<p>
+	If not empty, trigger if the commit message matches the regular expression.
+	The regular expression must be in the format that can be parsed by
+	<b>java.util.regex.Pattern</b>.
```
Comment (rsandell) [inline src/main/webapp/trigger/help-CommitMessageContains.html:7]:
```suggestion
	<a href="https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html#sum">Java Regex.</a>	
```
Or maybe some better wording. But the link is good to have for users not well versed in the Java platform.

**Review by rsandell — CHANGES_REQUESTED** [review]
Sorry for the delay in reviewing.

File: src/main/webapp/trigger/help-CommitMessageContains_ja.html (hunk 1 above)
Comment (bkhouri) [reply src/main/webapp/trigger/help-CommitMessageContains_ja.html:1]: I will remove

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/events/PluginPatchsetCreatedEvent.java (hunk 3 above)
Comment (bkhouri) [reply src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/events/PluginPatchsetCreatedEvent.java:224]: I shall update

File: src/main/webapp/trigger/help-CommitMessageContains.html (hunk 5 above)
Comment (bkhouri) [reply src/main/webapp/trigger/help-CommitMessageContains.html:7]: Makes sense. I was just following previous wording.

File: src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/event/PluginPatchsetCreatedEventTest.java (hunk 2 above)
Comment (bkhouri) [reply src/test/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/event/PluginPatchsetCreatedEventTest.java:94]: updated test name

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/events/PluginPatchsetCreatedEvent.java (hunk 4 above)
Comment (bkhouri) [reply src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/events/PluginPatchsetCreatedEvent.java:225]: I will update

**Comment by bkhouri** [comment]
> Sorry for the delay in reviewing.

@rsandell : Better late than never.   I believe all your concerns have been addressed. I will let you "Resolve conversation" if you are happy with my changes.

Also, do we need to explicitly make code changes to have this PR changes supported in pipeline syntax?

**Comment by bkhouri** [comment]
@rsandell I believe your concerns have been addressed. can you please have another look?  I force pushed a new change on top of `master` about 2 weeks ago.

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/events/PluginPatchsetCreatedEvent.java (hunk 3 above)
Comment (bkhouri) [reply src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/events/PluginPatchsetCreatedEvent.java]: this has been updated

File: src/main/webapp/trigger/help-CommitMessageContains.html (hunk 5 above)
Comment (bkhouri) [reply src/main/webapp/trigger/help-CommitMessageContains.html]: This has been updated.

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/events/PluginPatchsetCreatedEvent.java [hunk 6]
```diff
@@ -50,6 +53,10 @@
     private boolean excludeNoCodeChange = false;
     private boolean excludePrivateState = false;
     private boolean excludeWipState = false;
+    private String commitMessageContainsRegEx = "";
+
+    private transient Pattern commitMessagePattern = null;
+    private boolean forceUpdatePattern = true;
```
Comment (rsandell) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/events/PluginPatchsetCreatedEvent.java:59]:
```suggestion
    
```

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/events/PluginPatchsetCreatedEvent.java [hunk 7]
```diff
@@ -116,15 +129,28 @@ public void setExcludePrivateState(boolean excludePrivateState) {
...
 
+    /**
+     * Setter for commitMessageContains.
+     *
+     * @param commitMessageContainsRegEx Trigger if this text is included in the commit message
+     */
+    @DataBoundSetter
+    public void setCommitMessageContainsRegEx(String commitMessageContainsRegEx) {
+        this.commitMessageContainsRegEx = commitMessageContainsRegEx;
+        this.forceUpdatePattern = true;
```
Comment (rsandell) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/events/PluginPatchsetCreatedEvent.java:148]:
```suggestion
        this.commitMessagePattern = null;
```

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/events/PluginPatchsetCreatedEvent.java [hunk 8]
```diff
@@ -203,6 +243,15 @@ public boolean shouldTriggerOn(GerritTriggeredEvent event) {
         if (excludeWipState && ((PatchsetCreated)event).getChange().isWip()) {
             return false;
         }
+        if (StringUtils.isNotEmpty(commitMessageContainsRegEx)) {
+            if (null == commitMessagePattern || forceUpdatePattern) {
```
Comment (rsandell) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/events/PluginPatchsetCreatedEvent.java:247]:
```suggestion
            if (commitMessagePattern == null) {
```

**Review by rsandell — APPROVED** [review]
Looks good, I'll commit my own small suggestions before merging (just need to check that it compiles)

Next time, please don't force push amended commits, as I guess you are used to in Gerrit. In GitHub PRs it makes it really hard to review as I can't see the difference of what you've changed since I last reviewed and I have to do a full review again.

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/events/PluginPatchsetCreatedEvent.java [hunk 9]
```diff
@@ -203,6 +243,15 @@ public boolean shouldTriggerOn(GerritTriggeredEvent event) {
         if (excludeWipState && ((PatchsetCreated)event).getChange().isWip()) {
             return false;
         }
+        if (StringUtils.isNotEmpty(commitMessageContainsRegEx)) {
+            if (commitMessagePattern == null) {
+                commitMessagePattern = Pattern.compile(
+                        this.commitMessageContainsRegEx, Pattern.DOTALL | Pattern.MULTILINE);
+                forceUpdatePattern = false;
```
Comment (rsandell) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/hudsontrigger/events/PluginPatchsetCreatedEvent.java:250]:
```suggestion

```

**Pull request merged**